File: oxidize-python/oxidize_python/core/tokenizer/wordpiece.py
from __future__ import annotations

from dataclasses import dataclass, field

from oxidize_python.core.tokenizer.tokenizer import EncodeOptions, SpecialTokens


@dataclass
class WordPieceTokenizer:
    vocab: dict[str, int] = field(default_factory=dict)
    reverse: dict[int, str] = field(default_factory=dict)
    special: SpecialTokens = field(default_factory=SpecialTokens)
    unk_token: str = "[UNK]"
    max_input_chars: int = 100

    def __post_init__(self) -> None:
        if not self.reverse and self.vocab:
            self.reverse = {v: k for k, v in self.vocab.items()}
        if self.unk_token in self.vocab:
            self.special.unknown = self.vocab[self.unk_token]
        elif self.vocab:
            self.unk_token = next(iter(self.vocab))

    def name(self) -> str:
        return "wordpiece"

    def special_tokens(self) -> SpecialTokens:
        return self.special

    def vocab_size(self) -> int:
        return len(self.vocab)

    def encode(self, text: str, opts: EncodeOptions) -> list[int]:
        if not text:
            return []
        out: list[int] = []
        for word in _tokenize_basic(text):
            out.extend(self._tokenize_word(word))
        return out

    def _tokenize_word(self, word: str) -> list[int]:
        if len(word) > self.max_input_chars:
            return [self.special.unknown]
        pieces: list[str] = []
        start = 0
        while start < len(word):
            end = len(word)
            cur = ""
            ok = False
            while end > start:
                sub = word[start:end]
                if start > 0:
                    sub = "##" + sub
                if sub in self.vocab:
                    cur, ok = sub, True
                    break
                end -= 1
            if not ok:
                return [self.special.unknown]
            pieces.append(cur)
            start = end
        return [self.vocab.get(p, self.special.unknown) for p in pieces]

    def decode(self, tokens: list[int]) -> str:
        sb = []
        for tid in tokens:
            if tid in (self.special.cls, self.special.separator, self.special.pad):
                continue
            piece = self.reverse.get(tid, "")
            sb.append(piece.removeprefix("##"))
        return "".join(sb)

    def decode_skip_special(self, tokens: list[int]) -> str:
        return self.decode(tokens)


def _tokenize_basic(text: str) -> list[str]:
    words: list[str] = []
    current: list[str] = []
    for r in text:
        if r.isalnum():
            current.append(r)
        else:
            if current:
                words.append("".join(current))
                current = []
    if current:
        words.append("".join(current))
    return words

File: oxidize-golang/core/tokenizer/wordpiece.go
package tokenizer

import (
	"strings"
)

// WordPieceTokenizer is a WordPiece tokenizer (BERT family).
type WordPieceTokenizer struct {
	vocab      map[string]uint32
	reverse    map[uint32]string
	special    SpecialTokens
	unkToken   string
	maxInputChars int
}

// NewWordPieceTokenizer constructs a WordPiece tokenizer.
func NewWordPieceTokenizer(pieces []string, special SpecialTokens) *WordPieceTokenizer {
	rev := make(map[uint32]string, len(pieces))
	vocab := make(map[string]uint32, len(pieces))
	for i, p := range pieces {
		vocab[p] = uint32(i)
		rev[uint32(i)] = p
	}
	unk := "[UNK]"
	if id, ok := vocab[unk]; ok {
		special.Unknown = id
	} else {
		unk = pieces[0]
	}
	return &WordPieceTokenizer{
		vocab: vocab, reverse: rev, special: special,
		unkToken: unk, maxInputChars: 100,
	}
}

func (w *WordPieceTokenizer) Name() string { return "wordpiece" }
func (w *WordPieceTokenizer) SpecialTokens() SpecialTokens { return w.special }
func (w *WordPieceTokenizer) VocabSize() int               { return len(w.vocab) }

func (w *WordPieceTokenizer) Encode(text string, opts EncodeOptions) ([]uint32, error) {
	if text == "" {
		return []uint32{}, nil
	}
	words := tokenizeBasic(text)
	out := make([]uint32, 0, len(words))
	for _, word := range words {
		subTokens := w.tokenizeWord(word)
		out = append(out, subTokens...)
	}
	return out, nil
}

func (w *WordPieceTokenizer) tokenizeWord(word string) []uint32 {
	if len(word) > w.maxInputChars {
		return []uint32{w.special.Unknown}
	}
	pieces := make([]string, 0)
	start := 0
	for start < len(word) {
		end := len(word)
		var cur string
		ok := false
		for end > start {
			sub := word[start:end]
			if start > 0 {
				sub = "##" + sub
			}
			if _, found := w.vocab[sub]; found {
				cur = sub
				ok = true
				break
			}
			end--
		}
		if !ok {
			return []uint32{w.special.Unknown}
		}
		pieces = append(pieces, cur)
		start = end
	}
	out := make([]uint32, 0, len(pieces))
	for _, p := range pieces {
		if id, ok := w.vocab[p]; ok {
			out = append(out, id)
		} else {
			out = append(out, w.special.Unknown)
		}
	}
	return out
}

func (w *WordPieceTokenizer) Decode(tokens []uint32) (string, error) {
	var sb strings.Builder
	for _, id := range tokens {
		if id == w.special.CLS || id == w.special.Separator || id == w.special.Pad {
			continue
		}
		piece := w.reverse[id]
		piece = strings.TrimPrefix(piece, "##")
		sb.WriteString(piece)
	}
	return sb.String(), nil
}

func (w *WordPieceTokenizer) DecodeSkipSpecial(tokens []uint32) (string, error) {
	return w.Decode(tokens)
}

func tokenizeBasic(text string) []string {
	var words []string
	current := strings.Builder{}
	for _, r := range text {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			current.WriteRune(r)
		} else {
			if current.Len() > 0 {
				words = append(words, current.String())
				current.Reset()
			}
		}
	}
	if current.Len() > 0 {
		words = append(words, current.String())
	}
	return words
}
